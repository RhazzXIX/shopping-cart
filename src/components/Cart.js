import { useEffect, useState } from "react";
import del from "../images/del.svg";
import "../styles/Cart.css";

const Cart = (props) => {
  const { products, deleteProduct } = props;
  const [totalCost, setTotalCost] = useState(0);
  const getTotalCost = () => {
    let addedCost = 0;
    products.forEach((product) => {
      const productCost = product.price * product.qty;
      addedCost += productCost;
    });
    setTotalCost(addedCost);
  };

  useEffect(() => {
    getTotalCost();
  }, [products]);

  return (
    <section id="cart-page">
      <header>
        <h4>Cart</h4>
        <h5>Items:</h5>
        <h5>Costs:</h5>
      </header>
      {products.length !== 0 ? (
        <ul>
          {products.map((product) => {
            return (
              <li key={product.id}>
                <h4>{product.title}</h4>
                <img src={product.images[0]} alt={product.title} />
                <p>
                  $ {product.price}.00 x {product.qty}
                </p>
                <p>$ {product.price * product.qty}.00</p>
                <button onClick={deleteProduct}>
                  <img src={del} alt="delete" />
                </button>
              </li>
            );
          })}
        </ul>
      ) : (
        <p>No items on the cart</p>
      )}
      <footer>
        <h4>Total Cost: $ {totalCost}.00</h4>
      </footer>
    </section>
  );
};

export default Cart;
