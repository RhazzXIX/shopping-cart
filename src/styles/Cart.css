section#cart-page {
  position: absolute;
  right: 30px;
  top: 44px;
  width: 380px;
  background-color: white;
  padding: 18px;
  box-shadow: 0 0 4px black;
  border-radius: 8px;
  min-height: 220px;
  display: grid;
  grid-template-rows: auto 1fr auto;
}

section#cart-page header {
  display: grid;
  grid-template-rows: repeat(2 auto);
  grid-template-columns: 1fr 1fr;
  margin-bottom: 12px;
  row-gap: 6px;
}

section#cart-page header h5 {
  grid-row: 2 / 3;
}

section#cart-page header h5:last-child {
  justify-self: center;
}

section#cart-page ul li {
  display: grid;
  grid-template-rows: auto 1fr;
  grid-template-columns: repeat(3, auto);
  gap: 4px;
  border-radius: 12px;
  margin-bottom: 8px;
  padding: 8px;
}

section#cart-page > p {
  margin: 18px;
  margin-top: 26px;
  font-size: 1.2rem;
}

section#cart-page ul li:hover {
  box-shadow: 0 0 4px black;
}

section#cart-page li > img {
  width: auto;
  height: auto;
  max-height: 100px;
  max-width: 100px;
  grid-row: 2 / 3;
  border-radius: 8px;
}

section#cart-page li > p {
  grid-row: 2 / 3;
  align-self: center;
}

section#cart-page button {
  width: 20px;
  height: 20px;
  grid-column: 3 / 4;
  justify-self: center;
}

section#cart-page footer h4 {
  position: relative;
  left: 18%;
}
