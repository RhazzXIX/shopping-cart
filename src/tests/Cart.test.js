import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Cart from "../components/Cart";

const user = userEvent.setup();

const products = [];

const delProd = () => {
  products.pop();
};

const product = {
  title: "Black Motorbike",
  id: "214",
  price: 569,
  images: ["https://i.dummyjson.com/data/products/91/1.jpg"],
  qty: 2,
  description:
    "Engine Type:Wet sump, Single Cylinder, Four Stroke, Two Valves, Air Cooled with SOHC (Single Over Head Cam) Chain Drive Bore & Stroke:47.0 x 49.5 MM",
};

const product2 = {
  title: "iPhone X",
  id: "456",
  price: 899,
  images: ["https://i.dummyjson.com/data/products/2/1.jpg"],
  qty: 1,
  description:
    "SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip",
};

describe("Cart component", () => {
  beforeEach(() => products.push(product));
  it("Gets rendered", () => {
    const { container } = render(<Cart products={products} />);

    expect(container).toMatchSnapshot();
  });

  it("Can delete products", async () => {
    const { rerender } = render(
      <Cart products={products} deleteProduct={delProd} />
    );
    const delBtn = screen.getByRole("button");
    expect(screen.getByRole("listitem")).toBeInTheDocument();

    await user.click(delBtn);
    rerender(<Cart products={products} deleteProduct={delProd} />);
    expect(screen.queryByRole("listitem")).not.toBeInTheDocument();
  });

  it("Shows the total cost of all the products", () => {
    const { rerender } = render(
      <Cart products={products} deleteProduct={delProd} />
    );
    expect(
      screen.queryByRole("heading", { name: /total/i }).textContent
    ).toMatch("Total Cost: $ 1138.00");

    const addedProducts = [...products, product2];
    rerender(<Cart products={addedProducts} deleteProduct={delProd} />);
    expect(
      screen.queryByRole("heading", { name: /total/i }).textContent
    ).toMatch("Total Cost: $ 2037.00");
  });
  afterEach(() => products.splice(0));
});
