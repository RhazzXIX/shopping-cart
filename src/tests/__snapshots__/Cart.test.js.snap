// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Gets rendered 1`] = `
<div>
  <section
    id="cart-page"
  >
    <header>
      <h4>
        Cart
      </h4>
      <h5>
        Items:
      </h5>
      <h5>
        Costs:
      </h5>
    </header>
    <ul>
      <li>
        <h4>
          Black Motorbike
        </h4>
        <img
          alt="Black Motorbike"
          src="https://i.dummyjson.com/data/products/91/1.jpg"
        />
        <p>
          $ 
          569
          .00 x 
          2
        </p>
        <p>
          $ 
          1138
          .00
        </p>
        <button>
          <img
            alt="delete"
            src="del.svg"
          />
        </button>
      </li>
    </ul>
    <footer>
      <h4>
        Total Cost: $ 
        1138
        .00
      </h4>
    </footer>
  </section>
</div>
`;
